
## Introduction

Pipeline for RNASeq analysis, including RNA-Seq mapping and DEG analysis.

## RNA-Seq analysis

Please install STAR_2.5.3a first, then build the reference index. Also download the required annotation files. You can check the detailed mannual here: https://hbctraining.github.io/Intro-to-rnaseq-hpc-O2/lessons/03_alignment.html.

### Step 1: Mapping to genome

STAR have two modes to align the reads, and we prefer the two-pass mode. All detailed parameters are included in the STAR.sh. After mapping, there are two output files: one mapped to the transcriptome (for different expression genes analysis); the other one is mapped to whole genome (for alternative splicing, or SNV calling purpose) 

> STAR.sh TEST_R1.fastq.gz TEST_R2.fastq.gz TEST

### Step 2 and 3 : Qualitification with RSEM

After getting the bam mapped to transcriptome, apply quantificatiton with RSEM. 

> RSEM.sh Test_Aligned.toTranscriptome.out.bam TEST

Then you need to combine all RSEM output together. The results for next step is "gene.results"

> rsem-generate-data-matrix 1.genes.results 2.genes.results  > RSEM_transcript.genes.result

#### Step 4: Apply DEG analysis with DESeq2

Included is an example for Different expression analysis. The detailed mannual for DESeq2 is strongly suggested before running any analysis: http://bioconductor.org/packages/devel/bioc/vignettes/DESeq2/inst/doc/DESeq2.html

Basically, you need to prepare the input generated by RSEM, samples information, as well as your analysis plan. Please check S4_DEG.R for more details.

The Rscript output includes several vocalno plots. 

### Step 5: Apply gene set enrichment analysis. 

After getting the DEG list, the next goal is to identify which pathway is enriched. 

Goal of enrichment analysis: 

- Difference in a set of genes have more insight than individual gene
- Function annotation overrepresented in your gene list
- Find relationship between genes

Two strategy: 
DAVID: Use a subset of genes and check if these are functionally enriched

- Over-representation analysis (ORA), for example different express gene (DE)
- Need to choose threshold or cutoff, select <1000 genes
- Method: Fisherâ€™s Exact Test, Hypergeometric test, or Chiq Test

GSEA: Use all the genes and check if these are functionally enriched
- No need to choose a cutoff
- Genes are ranked using some metric
- Functional class scoring (FCS)

The script S5_GSEA.R is an example for running fGSEA package with the second strategy. 
